// @slidev/types/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});var __require = typeof require !== "undefined" ? require : (x) => {
  throw new Error('Dynamic require of "' + x + '" is not supported');
};

// ../../node_modules/.pnpm/tsup@5.1.0_typescript@4.4.3/node_modules/tsup/assets/cjs_shims.js
var importMetaUrlShim = typeof document === "undefined" ? new (__require("url")).URL("file:" + __filename).href : document.currentScript && document.currentScript.src || new URL("main.js", document.baseURI).href;

// src/setups.ts
function defineShikiSetup(fn) {
  return fn;
}
function defineWindiSetup(fn) {
  return fn;
}
function defineMonacoSetup(fn) {
  return fn;
}
function defineAppSetup(fn) {
  return fn;
}
function defineMermaidSetup(fn) {
  return fn;
}
function defineKatexSetup(fn) {
  return fn;
}
function defineShortcutsSetup(fn) {
  return fn;
}








exports.defineAppSetup = defineAppSetup; exports.defineKatexSetup = defineKatexSetup; exports.defineMermaidSetup = defineMermaidSetup; exports.defineMonacoSetup = defineMonacoSetup; exports.defineShikiSetup = defineShikiSetup; exports.defineShortcutsSetup = defineShortcutsSetup; exports.defineWindiSetup = defineWindiSetup;
